### ISSUE TITLE
Unexpected behavior when checkForUpdates has requirements

### ISSUE THREAD

**Issue by zenled** [infinum/iOS-prince-of-versions#35]
I have noticed some unexpected behavior when the `checkForUpdates` method needs to check for requirements.

Prince of Versions version: 4.0.0
iOS devices:  Simulator iOS 14.5; iPhone SE2 iOS 14.5
The App that was used for testing had a version set to 1.0

## Requirement check closure is called multiple times

The requirement check closure is called multiple times even if there is only one version entry.

```json
{
   "ios":[
      {
         "required_version":"1.2.4",
         "last_version_available":"1.8.0",
         "notify_last_version_frequency":"ALWAYS",
         "requirements":{
            "required_os_version":"8.0.0",
            "region":"hr",
            "bluetooth":"5.0"
         }
      }
   ]
}
```

```swift
let url = URL(string: "https://pastebin.com/raw/cS1LAXZT")!

let options = PoVRequestOptions()

options.addRequirement(key: "region"){(value) -> Bool in
   return true // breakpoint A
}

options.addRequirement(key: "bluetooth"){(value) -> Bool in
   return true // breakpoint B
}

PrinceOfVersions.checkForUpdates(from: url, options: options){ response in
    switch response.result {
    case .success(let updateResultData):
        print("Update version: \(updateResultData.updateVersion)") // breakpoint C
    case .failure:
        break
    }
}
```

I have set breakpoints `A`, `B`, and `C`.
I would expect the breakpoints to be triggered in the following order: `A`, `B`, `C`
The actual order is: `A`, `B`, `A`, `B`, `C`, `A`, `B`, `A`, `B`  or sometimes `B`, `A`, `B`, `A`, `C`, `B`, `A`, `B`, `A`

## `checkForUpdates` fails if `required_os_version` is missing from JSON 

If requirements in JSON do not include `required_os_version` the `checkForUpdates` will complete with failure.

```json
{
   "ios":[
      {
         "required_version":"1.2.4",
         "last_version_available":"1.8.0",
         "notify_last_version_frequency":"ALWAYS",
         "requirements":{
            "region":"hr",
            "bluetooth":"5.0"
         }
      }
   ]
}
```

```swift
let url = URL(string: "https://pastebin.com/raw/qtUwSvgR")!

let options = PoVRequestOptions()

options.addRequirement(key: "region"){(value) -> Bool in
   return true
}

options.addRequirement(key: "bluetooth"){(value) -> Bool in
   return true
}

PrinceOfVersions.checkForUpdates(from: url, options: options){ response in
    switch response.result {
    case .success(let updateResultData):
        print("Update version: \(updateResultData.updateVersion)")
    case .failure:
        break
    }
}
```

The `checkForUpdates` will complete with failure `(PrinceOfVersions.PoVError) dataNotFound`. The requirement check closures will not be called.
I would expect the `checkForUpdates` to call the requirement check closures and complete with success with the `UpdateStatus.requiredUpdateNeeded`.

## Failure if no requirements are satisfied

If none of the requirement check closures returns `true`, the `checkForUpdates` completes with failure.

```json
{
   "ios":[
      {
         "required_version":"1.2.4",
         "last_version_available":"1.8.0",
         "notify_last_version_frequency":"ALWAYS",
         "requirements":{
            "required_os_version":"8.0.0",
            "region":"hr",
            "bluetooth":"5.0"
         }
      }
   ]
}
```

```swift
let url = URL(string: "https://pastebin.com/raw/cS1LAXZT")!

let options = PoVRequestOptions()

options.addRequirement(key: "region"){(value) -> Bool in
   return false
}

options.addRequirement(key: "bluetooth"){(value) -> Bool in
   return false
}

PrinceOfVersions.checkForUpdates(from: url, options: options){ response in
    switch response.result {
    case .success(let updateResultData):
        PoVError
        print("Update version: \(updateResultData.updateVersion)")
    case .failure:
        break
    }
}
```

The `checkForUpdates` will complete with failure `(PrinceOfVersions.PoVError) requirementsNotSatisfied`.
Is this expected behavior? 
I would expect the `checkForUpdates` to complete with success with `UpdateStatus.noUpdateAvailable`

**Comment by ivanamrsic** [comment]
Hello @zenled 👋🏼  

First of all, I want to thank you for your very detailed report. It was very helpful when investigating issues you've reported. So, let's begin :)

**1. Requirement check closure is called multiple times**

This seems like an issue on our side and we will definitely look into it. The expected behavior is that each requirement check would be called exactly once for each configuration defined in JSON.  What this means exactly, is that, if there are multiple configurations defined in JSON, requirements checks will be done for every one of them so you can expect multiple calls of each `checkRequirement` method.

**2. `checkForUpdates` fails if `required_os_version` is missing from JSON**

Thank you for pointing this out. After going through our documentation and syncing with the Android team, we established that there are some discrepancies about this behavior. Your expectations are right, it is not mandatory for a user to define the `required_os_version` parameter in JSON. 

This, alongside the first issue, will be fixed in the next PR and documentation will be improved to make things clearer. :)

**3. Failure if no requirements are satisfied**

Regarding this issue, I think it is kinda debatable. Before further actions, I would like you to consider this two scenarios: 

* JSON with not one configuration that satisfies requirements (your case)
```
{
   "ios":[
      {
         "required_version":"1.2.4",
         "last_version_available":"1.8.0",
         "notify_last_version_frequency":"ALWAYS",
         "requirements":{
            "required_os_version":"8.0.0",
            "region":"hr",
            "bluetooth":"5.0"
         }
      }
   ]
}
```
* JSON where the configuration requirements are satisfied, however, the user has updated the app and he has no updates available
```
{
   "ios":[
      {
         "required_version":"1.2.4",
         "last_version_available":"1.8.0",
         "notify_last_version_frequency":"ALWAYS",
         "requirements":{
            "required_os_version":"7.0.0"
         }
      }
   ]
}
```

Would you consider this both equally as valuable as a developer? I understand it looks equal to the end-user, however, developers and app-owners should have the possibility to respond to these cases differently. This is the reasoning behind going with error `requirementsNotSatisfied` when requirements are indeed not satisfied for any configuration, and not with success option `UpdateStatus.noUpdateAvailable` .

Please let me know what you think and does this answers your questions. Also, if you have any more questions, be sure to reach out. 

Regards,
Ivana

**Comment by zenled** [comment]
Thank you for your answer.

Regarding **Failure if no requirements are satisfied**
I see the value for distinguishing between the two cases.
I guess I got confused by [this](https://github.com/infinum/iOS-prince-of-versions/blob/master/JSON.md#requirements) quote (`If there is not even one configuration that satisfies all requirements (including required_os_version), library will set updateStatus value to UpdateStatus.noUpdateAvailable`).

**Comment by ivanamrsic** [comment]
@zenled 

Nice catch! I'll change it accordingly :)

**Comment by ivanamrsic** [comment]
All problems listed in this issue are being fixed in #37 so I'm closing this issue.
